feat: Generate email workflow payload at time of sending

joeauyeung opened the pull request:
# Stacked on #24944

## What does this PR do?

https://www.loom.com/share/31fa7e381e7e409fa2f98984ef3a7fa3

This PR changes how email workflow content is generated for `BEFORE_EVENT` and `AFTER_EVENT` triggers. Instead of building the full email content at schedule time (eager), the system now stores only `bookingUid` and `workflowReminderId` in the tasker payload and generates the email content at send time (lazy).

**Why?** This ensures email reminders reflect the most up-to-date booking information (e.g., if a booking is rescheduled after the workflow was scheduled) and reduces the size of tasker payloads.

**Scope:** Only `BEFORE_EVENT` and `AFTER_EVENT` email workflow steps are affected. All other triggers (e.g., `EVENT_CANCELLED`, `RATING`, form triggers) continue to use the existing eager flow.



### Key Changes

**New Services:**
- `EmailWorkflowService` - Handles building email content (subject, body, ICS attachment, recipients) at send time
- `WorkflowService.scheduleLazyEmailWorkflow()` - Schedules lazy email tasks for BEFORE/AFTER event triggers
- `WorkflowService.processWorkflowScheduledDate()` - Computes scheduled dates based on trigger type
- `WorkflowService.generateCommonScheduleFunctionParams()` - Centralizes common scheduling parameters

**Repository Changes:**
- `WorkflowReminderRepository` - Converted from static to instance-based; added `create()` and `findByIdIncludeStepAndWorkflow()` methods
- `BookingSeatRepository` (new) - Fetches seat-specific attendee for seated events
- `BookingRepository` - Added `getBookingForCalEventBuilderFromUid()` method; extracted shared select statement

**Tasker Changes:**
- `sendWorkflowEmails` now accepts two payload types:
  - Eager: `{ to[], subject, html, ... }` (existing behavior)
  - Lazy: `{ bookingUid, workflowReminderId }` (new behavior)

**Type Centralization:**
- Moved `BookingInfo`, `AttendeeInBookingInfo`, `FormSubmissionData`, `ScheduleEmailReminderAction` to `ee/workflows/lib/types.ts`

## Visual Demo (For contributors especially)

N/A - Backend changes only, no UI modifications.

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. **BEFORE_EVENT email workflow:**
   - Create a workflow with BEFORE_EVENT trigger and EMAIL_ATTENDEE action
   - Create a booking that triggers the workflow
   - Reschedule the booking before the reminder fires
   - Verify the email contains the updated booking time

2. **AFTER_EVENT email workflow:**
   - Create a workflow with AFTER_EVENT trigger
   - Complete a booking and verify the email is sent with correct content

3. **Seated events:**
   - Create a seated event type with a BEFORE_EVENT email workflow
   - Book multiple seats
   - Verify each seat attendee receives their own email

4. **Non-affected triggers (regression):**
   - Verify EVENT_CANCELLED, RATING, and form-triggered workflows still work as expected

## Human Review Checklist

- [ ] Verify `WorkflowReminderRepository.create()` call in `WorkflowService.scheduleLazyEmailWorkflow` - the repository is now instance-based but appears to be called statically
- [ ] Confirm lazy generation only applies to BEFORE_EVENT/AFTER_EVENT email triggers
- [ ] Test backward compatibility with existing eager payloads

## Checklist


- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings






























<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Build email workflow content at send time instead of schedule time, using lazy tasks for before/after event email steps to keep messages accurate and reduce scheduling load. This matches the Linear requirement to generate the payload at time of sending.

- New Features
  - Lazy scheduling for BEFORE_EVENT and AFTER_EVENT email steps; task now carries bookingUid and workflowReminderId.
  - Task handler supports eager or lazy payloads; lazy path fetches booking by uid, builds the calendar event, then generates and sends email content.
  - EmailWorkflowService builds subject/body, ICS attachment, reply-to, and recipients (including seat-specific attendee).
  - WorkflowService computes scheduled dates and provides shared scheduling params.

- Refactors
  - Centralized types in ee/workflows/lib/types; simplified SMS manager.
  - WorkflowReminderRepository now instance-based with create and findByIdIncludeStepAndWorkflow.
  - Added BookingSeatRepository; BookingRepository exposes shared select and get-by-uid.
  - reminderScheduler and emailReminderManager now use services; emailProvider accepts nullable sendAt.

<sup>Written for commit 96a68e8a1eecea51b7d54f6ff9c3d32952abe4b3. Summary will update automatically on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

ThyMinimalDev committed:
wip

ThyMinimalDev committed:
wip

ThyMinimalDev committed:
feature: Booking Tasker without DI yet

ThyMinimalDev committed:
feature: Booking Tasker with DI

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

hbjORbj committed:
fix type check 1

hbjORbj committed:
fix type check 2

hbjORbj committed:
fix

ThyMinimalDev committed:
comment booking tasker for now

ThyMinimalDev committed:
fix: DI regularBookingService api v2

devin-ai-integration[bot] committed:
fix: convert trigger.dev SDK imports to dynamic imports to fix unit tests

The unit tests were failing because BookingEmailAndSmsTriggerTasker.ts had static imports of trigger files that depend on @trigger.dev/sdk. This caused Vitest to try to resolve the SDK at module load time, even though it should be optional.

Changed all imports in BookingEmailAndSmsTriggerTasker.ts from static to dynamic (using await import()) so the trigger files are only loaded when the tasker methods are actually called, not at module load time during tests.

This fixes the 'Failed to load url @trigger.dev/sdk' errors that were causing 28+ test failures.

Co-Authored-By: morgan@cal.com <morgan@cal.com>

hbjORbj committed:
fix unit tests

ThyMinimalDev committed:
keep inline smsAndEmailHandler.send calls

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: add team feature flag

ThyMinimalDev committed:
add satisfies ModuleLoader

ThyMinimalDev committed:
fix type check app flags

ThyMinimalDev committed:
move trigger in feature

hbjORbj committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: add trigger.dev prisma  generator

ThyMinimalDev committed:
fix: email app statuses

ThyMinimalDev committed:
fix: CalEvtBuilder unit test

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: improvements, schema, config, retry

ThyMinimalDev committed:
fixup! chore: improvements, schema, config, retry

ThyMinimalDev committed:
chore: cleanup code

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: cleanup code

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
chore: clean code and give full payload

ThyMinimalDev committed:
remove log

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
add booking notifications queue

ThyMinimalDev committed:
add attendee phone number for sms

ThyMinimalDev committed:
bump trigger to 4.1.0

ThyMinimalDev committed:
add missing booking seat data in attendee

ThyMinimalDev committed:
update config

ThyMinimalDev committed:
fix logger regular booking service

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: prisma as external deps of trigger

ThyMinimalDev committed:
fix yarn.lock

ThyMinimalDev committed:
revert change to example app booking page

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

devin-ai-integration[bot] committed:
fix: resolve circular dependencies and improve cold start performance in trigger tasks

- Convert BookingRepository import to type-only in CalendarEventBuilder.ts to eliminate circular dependency risk
- Convert EventNameObjectType, CalendarEvent, and JsonObject imports to type-only in BookingEmailAndSmsTaskService.ts
- Use dynamic imports in all trigger notification tasks (confirm, request, reschedule, rr-reschedule) to reduce cold start time
- Move heavy imports (BookingEmailSmsHandler, BookingRepository, prisma, TriggerDevLogger, BookingEmailAndSmsTaskService) inside run functions
- Eliminates module-level prisma import which violates repo guidelines and adds cold start overhead
- Reduces initial module dependency graph by deferring heavy imports (email templates, workflows, large repositories) until task execution

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

devin-ai-integration[bot] committed:
fix: improve cold start performance in reminderScheduler with dynamic imports

- Remove module-level prisma import (violates 'No prisma outside repositories' guideline)
- Use dynamic imports for UserRepository (1,168 lines) - only loaded when needed in EMAIL_ATTENDEE action
- Use dynamic imports for twilio provider (386 lines) - only loaded in cancelScheduledMessagesAndScheduleEmails
- Use dynamic imports for all manager functions by action type:
  - scheduleSMSReminder (387 lines) - loaded only for SMS actions
  - scheduleEmailReminder (459 lines) - loaded only for Email actions
  - scheduleWhatsappReminder (266 lines) - loaded only for WhatsApp actions
  - scheduleAIPhoneCall (478 lines) - loaded only for AI phone call actions
- Use dynamic imports for sendOrScheduleWorkflowEmails in cancelScheduledMessagesAndScheduleEmails
- Significantly reduces cold start time by deferring heavy module loading until execution paths need them
- Eliminates module-level prisma import that violated repository pattern guidelines

Co-Authored-By: morgan@cal.com <morgan@cal.com>

devin-ai-integration[bot] committed:
fix: improve cold start performance in BookingEmailSmsHandler with dynamic imports

- Remove module-level imports of all email-manager functions (653 LOC + 30+ email templates)
- Add dynamic imports in each method (_handleRescheduled, _handleRoundRobinRescheduled, _handleConfirmed, _handleRequested, handleAddGuests)
- Defer heavy email-manager loading until method execution
- Verified no circular dependencies between email-manager and bookings
- Significantly reduces cold start time for RegularBookingService and BookingEmailAndSmsTaskService

Co-Authored-By: morgan@cal.com <morgan@cal.com>

ThyMinimalDev committed:
fix: use dynamic imports

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
update yarn lock

ThyMinimalDev committed:
code review

ThyMinimalDev committed:
trigger config project ref in env

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
update yarn lock

ThyMinimalDev committed:
add .env.example trigger variables

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
add .env.example trigger variables

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: cleanup error handling and loggin

ThyMinimalDev committed:
fix: trigger config from env

ThyMinimalDev committed:
Merge branch 'main' into async-tasker-booking-email-and-sms

ThyMinimalDev committed:
fix: small typo fix

ThyMinimalDev committed:
fix: ai review comments

ThyMinimalDev committed:
fix: ai review comments

ThyMinimalDev committed:
ai review

joeauyeung committed:
Add `create` on `WorkflowReminderRepository`

joeauyeung committed:
`sendWorkflowEmails` tasker to accept lazy payload

joeauyeung committed:
Add `scheduleLazyEmailWorkflow` to `WorkflowService

joeauyeung committed:
Process scheduled date in `scheduleLazyEmailWorkflow`

joeauyeung committed:
Type fixes

joeauyeung committed:
Use `WorkflowService` to schedule

joeauyeung committed:
Refactor `scheduleEmailReminderForEvt` to use
`WorkflowService.processWorkflowScheduledDate`

joeauyeung committed:
Pass seat reference to lazy scheduled workflow reminder

joeauyeung committed:
Refactor `WorkflowReminderRepository` to accept prisma as constructor

joeauyeung committed:
Abstract `FormSubmissionData` type

joeauyeung committed:
Abstract select statement and add get by uid to `BookingRepository`

joeauyeung committed:
Add `FormSubmissionData` type

joeauyeung committed:
Add `findByIdIncludeStepAndWorkflow` to `WorkflowReminderRepository`

joeauyeung committed:
Tasker payload to accept `workflowReminderId`

joeauyeung committed:
Create `BookingSeatRepository`

joeauyeung committed:
Write `workflowReminderId` to tasker payload

joeauyeung committed:
Add `generateCommonScheduleFunctionParams` to `WorkflowService`

joeauyeung committed:
Init

joeauyeung committed:
Use services in tasker

joeauyeung committed:
Abstract types

joeauyeung committed:
In reminderScheduler use workflowService to generate common params

joeauyeung committed:
Type fix

joeauyeung committed:
Return params from emailWorkflowService.generateParametersToBuildEmailWorkflowContent

joeauyeung committed:
Use emailWorkflowService to generate params

joeauyeung committed:
Abstract types

joeauyeung committed:
Generate email content and send in EmailWorkflowService

joeauyeung committed:
Move check to caller

joeauyeung committed:
Use EmailWorkflowService to generate email payload in
emailReminderManager

joeauyeung commented on packages/features/ee/workflows/lib/reminders/emailReminderManager.ts:
```diff
@@ -136,200 +116,30 @@ const scheduleEmailReminderForEvt = async (args: scheduleEmailReminderArgs & { e
+  });
+  const workflowReminderRepository = new WorkflowReminderRepository(prisma);
+  const bookingSeatRepository = new BookingSeatRepository(prisma);
+  const emailWorkflowService = new EmailWorkflowService(workflowReminderRepository, bookingSeatRepository);
```
DI setup. In the future we can look at creating a DI container for the `EmailWorkflowService`

joeauyeung commented on packages/features/ee/workflows/lib/reminders/reminderScheduler.ts:
```diff
@@ -7,32 +7,24 @@ import {
+import type { BookingInfo } from "../types";
 import type { ScheduleTextReminderAction } from "./smsReminderManager";
 
-export type FormSubmissionData = {
```
Extract type

joeauyeung commented on packages/features/ee/workflows/lib/service/EmailWorkflowService.ts:
```diff
@@ -0,0 +1,422 @@
+    const creditService = new CreditService();
+    const creditCheckFn = creditService.hasAvailableCredits.bind(creditService);
+
+    const commonScheduleFunctionParams = WorkflowService.generateCommonScheduleFunctionParams({
```
Generate the common params needed to send an email workflow

joeauyeung commented on packages/features/ee/workflows/lib/service/EmailWorkflowService.ts:
```diff
@@ -0,0 +1,422 @@
+      teamId: workflow.teamId ?? undefined,
+    });
+
+    const emailWorkflowContentParams = await this.generateParametersToBuildEmailWorkflowContent({
```
Generate the params to generate the email payload

joeauyeung commented on packages/features/ee/workflows/lib/service/WorkflowService.ts:
```diff
@@ -239,4 +244,124 @@ export class WorkflowService {
+      scheduledAt: scheduledDate,
+      referenceUid: workflowReminder.uuid,
+    });
+  }
```
In the future we should be using Trigger

joeauyeung commented on packages/features/ee/workflows/lib/service/WorkflowService.ts:
```diff
@@ -239,4 +244,124 @@ export class WorkflowService {
+      referenceUid: workflowReminder.uuid,
+    });
+  }
+  static processWorkflowScheduledDate({
```
Logic taken from `emailReminderManager`

ThyMinimalDev commented on packages/features/ee/workflows/lib/service/EmailWorkflowService.ts:
```diff
@@ -0,0 +1,422 @@
+      );
+      emailAttendeeSendToOverride = seatAttendee?.attendee.email || null;
+    }
+    const creditService = new CreditService();
```
dynamic import credit service

joeauyeung committed:
Fix initalizing repository

joeauyeung committed:
Use evt.videoCallData first before the booking metadata

joeauyeung committed:
Only get non-deleted references to build calendar event

joeauyeung committed:
Remove check for videoCallData.id

joeauyeung committed:
Merge with main

joeauyeung committed:
Dynamic import credit service

joeauyeung committed:
BookingRepository.getByUid to only return what we need from attendee

joeauyeung commented on packages/features/CalendarEventBuilder.ts:
```diff
@@ -198,10 +198,10 @@ export class CalendarEventBuilder {
 
     // Video
-    if (videoCallData && videoCallData.id && videoCallData.url) {
+    if (videoCallData && videoCallData.url) {
```
Not every video reference has a `meetingId` ex Meet

joeauyeung committed:
Type fixes

devin-ai-integration[bot] committed:
test: Add comprehensive tests for lazy email workflow generation and fix broken tests

- Fix prisma mocks in sms-manager.test.ts and outOfOfficeCreateOrUpdate.handler.test.ts
  to export both 'default' and named 'prisma' exports
- Add EmailWorkflowService.test.ts with 4 tests for error handling paths
- Add sendWorkflowEmails.test.ts with 7 tests for schema validation and email sending
- Add tests to WorkflowService.test.ts for scheduleLazyEmailWorkflow,
  processWorkflowScheduledDate, and generateCommonScheduleFunctionParams methods

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

ThyMinimalDev commented on packages/features/ee/workflows/lib/service/EmailWorkflowService.ts:
```diff
@@ -0,0 +1,430 @@
+    });
+
+    await Promise.all(
+      emailWorkflowContentParams.sendTo.map((email) => {
```
@joeauyeung seems like a good comment from cubic

CarinaWolli reviewed (dismissed):
Code looks good. Tested it and everything worked as expected

emrysal reviewed (dismissed):
LGTM, tested through and happy.

emrysal committed:
Merge branch 'main' into tasker-generate-workflow-body

joeauyeung committed:
Write `seatReferenceUid`

joeauyeung committed:
Return promise

joeauyeung committed:
Fix log

joeauyeung committed:
Change to `Promise.allSettled`

joeauyeung committed:
Type fix

joeauyeung committed:
Fix failing test

devin-ai-integration[bot] committed:
fix: reorder workflow step checks to fix test failure

The test 'should throw error if workflow step not found on reminder' was failing because the code checked workflowStep.verifiedAt before checking if workflowStep exists. When workflowStep is null, this caused the error message to include 'undefined' instead of the expected workflow step id.

Fixed by reordering the checks:
1. First check if workflowStep exists
2. Then check if workflowStep.verifiedAt exists

Also updated the test expectation to match the correct error message.

Co-Authored-By: joe@cal.com <j.auyeung419@gmail.com>

joeauyeung committed:
Merge branch 'main' into tasker-generate-workflow-body

joeauyeung committed:
Merge branch 'main' into tasker-generate-workflow-body

emrysal reviewed (approved):
re-approved
